File: src/battle/state.ts
/**
 * Battle state.
 *
 * The whole fight is one serializable object: combatants, round number, RNG
 * state. Resolving a round is a function from (state, commands) to (state,
 * events) - the caller decides whether to mutate in place or keep the previous
 * state around for rollback. Because the RNG lives inside the state, saving
 * mid-battle and reloading reproduces the same rolls.
 */

import type { RngState } from '../core/rng.js';
import type { ActorState } from '../domain/actor.js';
import type { StatBlock } from '../domain/stats.js';
import type { Element } from '../domain/elements.js';

export type Side = 'party' | 'foe';

export interface Combatant {
  /** Unique within the battle. */
  id: string;
  side: Side;
  /** Formation slot; adjacency drives spread damage falloff. */
  slot: number;
  /** Party members carry a full actor state; enemies carry a stat block. */
  actor?: ActorState;
  enemyId?: string;
  name: string;
  innate: Element;
  hp: number;
  maxHp: number;
  aether: number;
  maxAether: number;
  /** Enemy-only cached stats; party stats derive from the actor each read. */
  enemyStats?: StatBlock;
  statuses: { statusId: string; remaining: number; sourceId?: string | null }[];
  defending: boolean;
  downed: boolean;
  /** Rounds until each move may be used again, keyed by art id. */
  cooldowns: Record<string, number>;
  tempModifiers: { modifier: import('../domain/stats.js').StatModifier; rounds: number }[];
}

export type BattlePhase = 'command' | 'resolving' | 'ended';

export interface BattleState {
  round: number;
  phase: BattlePhase;
  rng: RngState;
  combatants: Combatant[];
  outcome: 'victory' | 'defeat' | 'fled' | null;
  /** Accumulated rewards, filled in on victory. */
  rewards: { xp: number; coin: number; itemIds: string[] };
  /** Set when the encounter forbids fleeing. */
  noFlee: boolean;
}

export type BattleCommand =
  | { kind: 'attack'; combatantId: string; targetId: string }
  | { kind: 'art'; combatantId: string; artId: string; targetId?: string }
  | { kind: 'unleash'; combatantId: string; moteId: string; targetId?: string }
  | { kind: 'summon'; combatantId: string; summonId: string; targetId?: string }
  | { kind: 'defend'; combatantId: string }
  | { kind: 'item'; combatantId: string; itemId: string; targetId?: string }
  | { kind: 'flee'; combatantId: string };

export function livingCombatants(state: BattleState, side: Side): Combatant[] {
  return state.combatants.filter((c) => c.side === side && !c.downed);
}

export function findCombatant(state: BattleState, id: string): Combatant | undefined {
  return state.combatants.find((c) => c.id === id);
}

export function opposingSide(side: Side): Side {
  return side === 'party' ? 'foe' : 'party';
}

/** Slot distance, used for spread damage falloff. */
export function slotDistance(a: Combatant, b: Combatant): number {
  return Math.abs(a.slot - b.slot);
}

export function isBattleOver(state: BattleState): boolean {
  return livingCombatants(state, 'party').length === 0 || livingCombatants(state, 'foe').length === 0;
}

File: unity/Aetherlight.Core/Battle/State.cs
using System;
using System.Collections.Generic;
using Aetherlight.Core;
using Aetherlight.Domain;

namespace Aetherlight.Battle
{
    public enum Side { Party, Foe }

    public enum BattlePhase { Command, Resolving, Ended }

    public sealed class Combatant
    {
        /// <summary>Unique within the battle.</summary>
        public string Id = "";
        public Side Side;
        /// <summary>Formation slot; adjacency drives spread damage falloff.</summary>
        public int Slot;

        /// <summary>Party members carry a full actor state; enemies carry a stat block.</summary>
        public ActorState? Actor;
        public string? EnemyId;

        public string Name = "";
        public Element Innate;
        public double Hp;
        public double MaxHp;
        public double Aether;
        public double MaxAether;

        /// <summary>Enemy-only cached stats; party stats derive from the actor each read.</summary>
        public StatBlock? EnemyStats;

        public List<StatusInstance> Statuses = new List<StatusInstance>();
        public bool Defending;
        public bool Downed;

        /// <summary>Rounds until each move may be used again, keyed by art id.</summary>
        public Dictionary<string, int> Cooldowns = new Dictionary<string, int>();
        public List<TimedModifier> TempModifiers = new List<TimedModifier>();
    }

    public sealed class BattleRewards
    {
        public long Xp;
        public long Coin;
        public List<string> ItemIds = new List<string>();
    }

    public sealed class BattleState
    {
        public int Round = 1;
        public BattlePhase Phase = BattlePhase.Command;
        public RngState Rng;
        public List<Combatant> Combatants = new List<Combatant>();
        public BattleOutcome? Outcome;
        public BattleRewards Rewards = new BattleRewards();
        /// <summary>Set when the encounter forbids fleeing.</summary>
        public bool NoFlee;
    }

    public enum CommandKind { Attack, Art, Unleash, Summon, Defend, Item, Flee }

    public sealed class BattleCommand
    {
        public CommandKind Kind;
        public string CombatantId = "";
        public string? TargetId;
        public string? ArtId;
        public string? MoteId;
        public string? SummonId;
        public string? ItemId;

        public static BattleCommand Attack(string combatantId, string targetId) =>
            new BattleCommand { Kind = CommandKind.Attack, CombatantId = combatantId, TargetId = targetId };

        public static BattleCommand Art(string combatantId, string artId, string? targetId = null) =>
            new BattleCommand { Kind = CommandKind.Art, CombatantId = combatantId, ArtId = artId, TargetId = targetId };

        public static BattleCommand Unleash(string combatantId, string moteId, string? targetId = null) =>
            new BattleCommand { Kind = CommandKind.Unleash, CombatantId = combatantId, MoteId = moteId, TargetId = targetId };

        public static BattleCommand Summon(string combatantId, string summonId) =>
            new BattleCommand { Kind = CommandKind.Summon, CombatantId = combatantId, SummonId = summonId };

        public static BattleCommand Defend(string combatantId) =>
            new BattleCommand { Kind = CommandKind.Defend, CombatantId = combatantId };

        public static BattleCommand Flee(string combatantId) =>
            new BattleCommand { Kind = CommandKind.Flee, CombatantId = combatantId };
    }

    public static class Battles
    {
        public static List<Combatant> Living(BattleState state, Side side)
        {
            var living = new List<Combatant>();
            foreach (var combatant in state.Combatants)
            {
                if (combatant.Side == side && !combatant.Downed) living.Add(combatant);
            }
            return living;
        }

        public static Combatant? Find(BattleState state, string id) =>
            state.Combatants.Find(c => c.Id == id);

        public static Side Opposing(Side side) => side == Side.Party ? Side.Foe : Side.Party;

        /// <summary>Slot distance, used for spread damage falloff.</summary>
        public static int SlotDistance(Combatant a, Combatant b) => Math.Abs(a.Slot - b.Slot);

        public static bool IsOver(BattleState state) =>
            Living(state, Side.Party).Count == 0 || Living(state, Side.Foe).Count == 0;
    }
}
